deployer/billing/cost_importer.py

```diff
@@ -0,0 +1,185 @@
+        self.cluster = cluster
+
+    def get_costs(self, start_month, end_month):
+        rows = self._run_query(start_month, end_month)
+        return self.clean_query(rows)
+
+    def _run_query(self, start_month, end_month):
+        # Write doc string
+        url, s = get_cluster_prometheus(self.cluster.get("name"))
+        p = query.Prometheus(url, http=s)
+        prom_query = """sum(
+        kube_pod_container_resource_requests{resource="memory",unit="byte"}
+        ) by (namespace)
+        """
+        start = start_month.isoformat()
+        end = end_month.isoformat()
+        step = "24h"
+        rows = p.query_range(prom_query, start, end, step)
+        return rows
+
+    def clean_query(self, rows):
+        # This only handles single prometheus namespace labels rather than any general prometheus query.
+        # The regex should be for an RFC 1123 DNS_LABEL however we're only querying namespaces so strict
+        # validation not as necessary
+        # https://github.com/kubernetes/design-proposals-archive/blob/main/architecture/identifiers.md
+        rows.rename(
+            columns=lambda c: re.sub(
+                r".*=\"(?P<namespace>[a-zA-Z0-9-]+)\"}", r"\g<namespace>", c
+            ),
+            inplace=True,
+        )
+
+        rows.index.names = ["month"]
+        rows.fillna(0, inplace=True)
+        rows = rows.resample("MS", axis=0).sum()
+        # Calculate utilization
+        # rows = rows.div(rows.sum(axis=1), axis=0)
+        # Combined support
+        # rows["support_combined"] = rows["support"] + rows["kube-system"]
+        return rows
+
+
+def get_cluster_costs(cluster, start_month, end_month):
+    tenancy = cluster.get("tenancy", "dedicated")
+    if tenancy == "shared":
+        return get_shared_hub_costs(cluster, start_month, end_month)
+    elif tenancy == "dedicated":
+        return get_dedicated_cluster_costs(cluster, start_month, end_month)
+    return pd.DataFrame()
+    # return get_dedicated_cluster_costs(cluster, start_month, end_month)
+
+
+def get_dedicated_cluster_costs(cluster, start_month, end_month):
+    """Return monthly costs for a dedicated cluster for a range of months
+
+    Args:
+        cluster (dict): parsed cluster.yaml
+        start_month (datetime): Starting month
+        end_month (datetime): End month
+
+    Returns:
+        rows (dataframe of costs per project)
+    """
+    bq_importer = BigqueryCostImporter(cluster)
+    result = bq_importer.get_costs(start_month, end_month)
+    return result
+
+
+# TODO(pnasrat): wire in prometheus values
+def get_shared_hub_costs(cluster, start_month, end_month):
+    costs = get_dedicated_cluster_costs(cluster, start_month, end_month)
+    prom_importer = PrometheusUtilizationImporter(cluster)
+    utilization = prom_importer.get_costs(start_month, end_month)
+    totals = costs.merge(utilization, how="left", on="month")
+    totals.iloc[:, 2:].multiply(totals.iloc[:, 1].astype(float), axis=0)
```

Comment:
If I understand this correctly, this is the core of computing shared costs - we get these to be two pd dataframes and use pandas to do appropriate “math”. We can extend this in complexity in the future by adding appropriate extra columns, etc. I don’t know a lot of pandas, but the community is large enough that i believe we can make it work.


If my assumption of what this approach is is correct, then i am a big fan! I vastly prefer this to the lists & dict stuff i had written for dedicated clusters earlier!

i will provide a more detailed review monday, but at a general level, +1!

Reply:
Yes at a high level I'm trying to make it more portable going forward using pandas. It allows us to replicate the approach from the spreadsheets programatically and is extendible as  you say. Some of the slowgoing has been me trying to figure out good practices in libraries I'm less familiar with.

Reply:
@pnasrat awesome, i love the approach!!!!